src/main/java/com/atlassian/db/replica/api/AuroraPostgresLsnReplicaConsistency.java

```diff
@@ -0,0 +1,99 @@
+import com.atlassian.db.replica.spi.Cache;
+import com.atlassian.db.replica.spi.ReplicaConsistency;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * It remembers last write LSN (log sequence number) and compares it with LSN for subsequent reads to detect inconsistencies.
+ * If it cannot remember the LSN of last write, pessimistically assumes it's going to be inconsistent.
+ *
+ * @see <a href="https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database-monitoring.html">Monitoring Aurora PostgreSQL-based Aurora global databases</a>
+ * <p>
+ * It's an PostgreSQL-based Aurora database engine specific implementation.
+ */
+public final class AuroraPostgresLsnReplicaConsistency implements ReplicaConsistency {
+
+    private final Cache<Long> lastWrite;
+
+    public static final class Builder {
+        private Cache<Long> lastWrite = new MonotonicMemoryCache<>();
+
+        /**
+         * @param lastWrite remembers last write
+         */
+        public Builder cacheLastWrite(Cache<Long> lastWrite) {
+            this.lastWrite = lastWrite;
+            return this;
+        }
+
+        /**
+         * @return consistency assuming that LSN (log sequence number) is greater or equal to LSN for last write (if known)
+         */
+        public AuroraPostgresLsnReplicaConsistency build() {
+            return new AuroraPostgresLsnReplicaConsistency(lastWrite);
+        }
+    }
+
+    private AuroraPostgresLsnReplicaConsistency(Cache<Long> lastWrite) {
+        this.lastWrite = lastWrite;
+    }
+
+    @Override
+    public void write(Connection main) {
+        try {
+            lastWrite.put(queryMainDbLsn(main));
+        } catch (Exception e) {
+            lastWrite.reset();
+        }
+    }
+
+    @Override
+    public boolean isConsistent(Supplier<Connection> replica) {
+        Optional<Long> maybeLastWrite = lastWrite.get();
+        if (!maybeLastWrite.isPresent()) {
+            return false;
+        }
+
+        long lastRefresh;
+        try {
+            lastRefresh = queryReplicaDbLsn(replica.get());
+        } catch (Exception e) {
+            return false;
+        }
+
+        return lastRefresh >= maybeLastWrite.get();
+    }
+
+    /**
+     * @return LSN (log sequence number) for the main database (the highest LSN)
+     */
+    private long queryMainDbLsn(Connection connection) throws Exception {
+        return queryLsn(connection, "SELECT MAX(durable_lsn) AS lsn FROM aurora_global_db_instance_status();");
+    }
+
+    /**
+     * @return LSN (log sequence number) for the most outdated replica database (the lowest LSN)
+     */
+    private long queryReplicaDbLsn(Connection connection) throws Exception {
+        return queryLsn(connection, "SELECT MIN(durable_lsn) AS lsn FROM aurora_global_db_instance_status();");
```

Comment:
The `durable_lsn`:
> The LSN made durable in storage. 

If replica dies and gets restarted, it will have this LSN. But `DualConnection` doesn't care about durability, but visibility. So we should query `highest_lsn_rcvd`. Right?

Reply:
I didn't found good explanation of differences between fields, so this is my understanding:
- durable_lsn (it's persisted and served)
- highest_lsn_rcvd (it's received, not persisted and not served yet?)
- oldest_read_view_lsn (lsn during last read operation)
I have doubts whether `highest_lsn_rcvd` means that those changes are served to clients. @dagguh WDYT?

Reply:
`highest_lsn_rcvd ` in case of main db is always null:
<img width="1722" alt="Screenshot 2021-03-17 at 15 39 10" src="https://user-images.githubusercontent.com/77149142/111485461-eda61380-8736-11eb-88d9-96f4ead37d71.png">

Reply:
so if "durable" seems to be lower bound for "visible" then it's the safest option here.

Reply:
Make sense. Maybe we can use it instead of aggregation to distinguish between main and replicas.

Reply:
@wyrzyk Will it give us much performance gain (`WHERE hightest_lsn_rcsv` != NULL`) than aggregation with `MIN/MAX`?

Reply:
Interesting. Is there an alternative call which shows LSN? Maybe the Postgres function?

It's also interesting to see that replica `highest_lsn_rcvd` is bigger than its `durable_lsn`, which seems to confirm my assumption. In this case we can see that replica received the latest durable LSN from main.
Maybe main only has durable LSN because it doesn't "receive" LSN, it is sending it.

So here's my proposal:
* read `durable_lsn` from main
* read `highest_lsn_rcvd` from replica

Reply:
> Will it give us much performance gain (WHERE hightest_lsn_rcsv != NULL) than aggregation with MIN/MAX?

I have no idea.

Reply:
@dagguh Whole point of LSN usage is to avoid consistency errors. We are not sure whether `highest_lsn_rcvd ` will serve changes between `durable_lsn ` and `highest_lsn_rcvd` so it can introduce consistency issue. IMO we should use  `durable_lsn ` for both.

Reply:
Very good point. That's the safe way 👍

Reply:
So the current code is correct